### ISSUE TITLE
obs(model): expose immutable CUDA weight upload deltas for #9420

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#9754]
## Parent

#9420. This is the enabling instrumentation child required to finish its matched A100 acceptance receipt without putting a second implementation commit under the parent issue.

## Problem frame

Centrality: Enabling.

For operators validating exact Qwen3.8 cache restoration on fak-native CUDA; the problem is that the production path exposes neither truthful immutable-weight H2D call/byte accounting nor direct first-real-Step versus steady-decode timing; today test-only model recording and token-commit timestamps can only imply those costs; this is better because opt-in, backend-authored counters and direct forward timing make the parent acceptance falsifiable without changing model math; the witness is failure-first CUDA upload-cache tests plus an opt-in native-inference/decode-trace gateway response.

- P1: preserved — counters observe the exact fak-native CUDA and Step/StepBatchActive seams without changing model math, cache state, or output.
- P2: advanced — repeated upload work and first-hit decode setup become attributable instead of inferred from aggregate wall time.
- P3: preserved — accounting is monotonic, concurrency-safe, post-success, and omitted unless explicitly requested.
- P4: advanced — the parent 5+5 campaign can bind calls/bytes and direct forward timing beside cache identity in one machine-readable response.

## Current state

`internal/model/device_weight_residency.go` shares immutable handles across sessions, but counting its stage attempts would be untruthful because `internal/compute/cuda.go` can satisfy an upload from its pointer cache. CUDA cumulative H2D bytes include activations and KV and do not classify immutable weights. `NativeDecodeTrace` records token commits before the following `Session.Step`, so event gaps mix forward, sampling, and emit overhead rather than isolating the first real Step.

## Smallest working spine

1. At the actual CUDA upload-cache miss, increment immutable-weight upload calls/transfer bytes/resident bytes only after successful H2D completion; exclude activations and refused/cached uploads.
2. Expose a cumulative optional CUDA snapshot without widening the common Backend interface; keep transfer bytes distinct from resident bytes for F16, Q8/Q2 scale sidecars, and raw Q4/Q5/Q6.
3. Measure the forward call directly around serial `Session.Step` and batched `StepBatchActive`; attach the duration to the committed token that caused that forward, documenting batch wall time for active lanes.
4. Surface opt-in before/after CUDA counter deltas and per-token forward timings through the existing native receipt/decode trace production gateway wire; keep default wire and unprofiled execution unchanged.

## Done condition

- [ ] Same CUDA backend and host pointer uploaded twice records one actual immutable-weight H2D call; a new pointer adds one; activation uploads add zero.
- [ ] Call, transfer-byte, and resident-byte totals reflect the actual dtype layout and update only after successful copies.
- [ ] Exact cached-logits decode with at least three emitted tokens records event 1’s forward as the first real Step, event 2’s forward as steady decode, and no nonexistent final forward.
- [ ] Serial and batched timing semantics are explicit and outputs remain unchanged.
- [ ] An opt-in production gateway response carries request/campaign-usable upload deltas and forward timing; default wire omission remains unchanged.

## Gold-plating boundary

No CUDA kernel change, no paged-swap codec work, no llama.cpp fallback, no hardware number, no Remote-L3 work, and no mutation of #9420’s accepted Qwen3.8/A100 envelope.

## Likely files

- `internal/compute/cuda_backend_state.go`
- `internal/compute/cuda.go`
- focused CUDA-tagged tests
- `internal/agent/chat.go`
- `internal/agent/inkernel_decode.go`
- `internal/agent/inkernel_planner.go`
- focused agent/gateway receipt and trace tests

## Witness

Failure-first affected-package tests prove actual post-cache H2D accounting, exact direct Step timing on cached logits, request-visible JSON reachability, and default wire omission. Final parent acceptance remains the sanctioned sequential A100 5+5 artifact-bound run; cumulative deltas are interpreted only across that deliberately serialized campaign.

## Lane

compute+agent+gateway

## Work unit

leaf

## Completion standard

production

**Comment by anthony-chaudhary** [comment]
Claimed for direct bounded implementation.

Exact generation-1 leased scope:
- `internal/compute/cuda_backend_state.go`
- `internal/compute/cuda.go`
- `internal/compute/cuda_upload_observability_test.go`
- `internal/agent/chat.go`
- `internal/agent/inkernel_decode.go`
- `internal/agent/inkernel_planner.go`
- `internal/agent/inkernel_decode_trace_test.go`
- `internal/gateway/chat_decode_trace_test.go`
- `internal/gateway/native_inference_receipt_test.go`

Scope remains one coherent compute→agent→gateway observability spine. The CUDA backend will expose an optional concrete-method cumulative snapshot, so the common `compute.Backend` interface is unchanged. Accounting is post-H2D and cache-miss-only for cache-eligible immutable matrices; activation/classed uploads, cache hits, refused uploads, and empty/no-copy paths remain zero. F16 transfer bytes remain source-F32 H2D bytes while resident bytes are F16; quantized paths count codes plus scale sidecars; raw K-quant counts packed bytes.

Agent timing is enabled only by the existing decode-trace opt-in, measured immediately around `Session.Step` or the one `StepBatchActive` wall-time shared by active lanes, and attached to the causing committed token. The final token has no forward record. Existing native-inference opt-in will carry backend cumulative before/after/delta values; default wire remains absent. No production gateway file is needed because it already serializes the typed agent receipts.

Witness plan: failure-first CUDA-tagged accounting tests, serial/batched direct-forward tests, request-local backend delta tests, gateway JSON/default omission tests, focused/race and strict full affected packages, exact-path validation, then signed exact-path landing only on green. No kernel/math/paging/Remote-L3/fallback/hardware claim.

**Comment by anthony-chaudhary** [comment]
Implementation checkpoint — held before landing because the required CUDA-tagged accounting witness has not yet run on a device.

Exact owned dirty paths remain the nine paths in the claim comment. No scope expansion and no common `compute.Backend`, kernel/math, paging, Remote-L3, fallback, or execution-default change.

Green gates on current trunk:
- `go test ./internal/compute -count=1` — PASS (the unrelated `/bin/true` baseline blocker #9792 is fixed on trunk)
- `go test ./internal/agent -count=1` — PASS
- `go test ./internal/gateway -count=1` — PASS
- focused agent and gateway tests repeated — PASS
- focused `-race` agent and gateway tests repeated — PASS
- exact nine-path `fak validate --mine ...` — PASS

The new CUDA-tagged tests are `TestCUDAImmutableWeightUploadSnapshotCountsOnlySuccessfulCacheMisses` and `TestCUDAImmutableWeightUploadSnapshotUsesActualTransferAndResidentLayouts`. This macOS control point cannot link the CUDA build (`libfakcuda` absent), which is not a pass or hardware claim. The private-lab bridge preflight currently lacks its operator token/channel configuration, and the public-cloud probe reports stale credentials, so neither sanctioned device route is presently dispatchable from this process.

Next checkable operator step on an authenticated CUDA build node, using these held working-tree bytes: build `libfakcuda.a` for the node architecture, then run

```
go test -tags cuda ./internal/compute -run '^TestCUDAImmutableWeightUploadSnapshot' -count=1 -v
```

The output must show both tests actually ran and passed (not skipped). Only then rerun the strict affected gates, exact-path commit/push, post promotion/demotion/invalidation evidence, and close #9754. No commit or closure is being claimed while that device witness is missing.

**Comment by anthony-chaudhary** [comment]
Source observability spine split to #9794 after all sanctioned GPU routes and the repository runner inventory returned unavailable. #9794 owns the exact nine-file compute→agent→gateway counter/timing implementation and GPU-free/full-package/race proof. This parent retains only the real CUDA-tagged device accounting receipt and must remain open until a sanctioned GPU returns it.

**Comment by anthony-chaudhary** [comment]
Tracker reconciliation: the already-implemented source instrumentation spine has moved to child #9794, with the exact nine paths released from #9754 and reacquired under #9794. This parent remains open solely for the sanctioned real-device CUDA-tagged accounting witness described in the prior handoff; no fake-backend, ABI, compile, or source-level pass will be promoted into a device claim.

**Comment by anthony-chaudhary** [comment]
Completed as the instrumentation child of #9420, with the public production seam supplied by #9794 and hardening in pushed commit 504bf5a0c (internal/agent@r366+g504bf5a0c; internal/compute@r216+g504bf5a0c). Receipt semantics now cover successful immutable-weight H2D operations, actual transfer/resident bytes, matrix hit suppression, uncached rank-one copies, Q2 lifetime rooting, and per-real-Step timing. Focused committed-tip agent/gateway tests and pre-push graph gate are green. Hardware numbers are intentionally not claimed here; #9420 retains the 5-cold/5-hit A100 acceptance.
